File: smalivm/src/main/java/org/cf/smalivm/opcode/FillArrayDataOpFactory.java
package org.cf.smalivm.opcode;

import gnu.trove.map.TIntObjectMap;

import org.cf.smalivm.VirtualMachine;
import org.jf.dexlib2.builder.BuilderInstruction;
import org.jf.dexlib2.builder.MethodLocation;
import org.jf.dexlib2.iface.instruction.OffsetInstruction;
import org.jf.dexlib2.iface.instruction.formats.Instruction31t;

public class FillArrayDataOpFactory implements OpFactory {

    @Override
    public Op create(MethodLocation location, TIntObjectMap<MethodLocation> addressToLocation, VirtualMachine vm) {
        BuilderInstruction instruction = (BuilderInstruction) location.getInstruction();
        int address = instruction.getLocation().getCodeAddress();
        int returnAddress = address + instruction.getCodeUnits();
        MethodLocation returnLocation = addressToLocation.get(returnAddress);
        int branchOffset = ((OffsetInstruction) instruction).getCodeOffset();
        int childAddress = address + branchOffset;
        MethodLocation child = addressToLocation.get(childAddress);
        Instruction31t instr = (Instruction31t) location.getInstruction();
        int register = instr.getRegisterA();

        return new FillArrayDataOp(location, child, returnLocation, register);
    }
}

File: smalivm/src/main/java/org/cf/smalivm/opcode/FillArrayDataOpFactory.kt
package org.cf.smalivm.opcode

import gnu.trove.map.TIntObjectMap
import org.cf.smalivm.VirtualMachine
import org.jf.dexlib2.builder.BuilderInstruction
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.OffsetInstruction
import org.jf.dexlib2.iface.instruction.formats.Instruction31t

class FillArrayDataOpFactory : OpFactory {
    fun create(location: MethodLocation, addressToLocation: TIntObjectMap<MethodLocation>, vm: VirtualMachine?): Op {
        val instruction = location.instruction as BuilderInstruction?
        val address = instruction!!.location.codeAddress
        val returnAddress = address + instruction.codeUnits
        val returnLocation = addressToLocation[returnAddress]
        val branchOffset = (instruction as OffsetInstruction?)!!.codeOffset
        val childAddress = address + branchOffset
        val child = addressToLocation[childAddress]
        val instr = location.instruction as Instruction31t?
        val register = instr!!.registerA
        return FillArrayDataOp(location, child, returnLocation, register)
    }
}
